### PR TITLE
Systematics can be associated with a group of EDs.

### PR DESCRIPTION
**Pull request by BillyLiggins**
The SystematicManager now stores its systematics in map of vectors.
Each vector is referred to as a group and is keyed by the name of that
group. Systematics are added to a group through the BinnedNLLH (lh)
like:

    lh.AddSystematic(Systematic, "groupName");

The order in which the systematic is added to a group is the order in
which it is applied. If a group name is not specified, the systematic is
added the global ("") group which is then *always* applied to all EDs
before any other groups.

A ED is then associated with a vector of groups that will be applied to
that ED in that order:

    std::vector<std::string> appliedSys;
    appliedSys.push_back("GroupA");
    appliedSys.push_back("GroupB");
    lh.AddDist(ED,"appliedSys")

Global and group bases systematics may be applied together.

See examples/systematicFit.cpp for example fit.

Other changes

\#\#\# Details

There have been some changes to the to the previous PR.

\#\#\#\#\# Aim \#\#\#\#\#
We want to be able to add systematics to specific groups e.g.
"alphaEnergyRes", which can then be chained to each other and then
associated with a distribution.  However we also want to retain the
previous behaviour of adding global systematics to the lh which then
acts on all distributions. The problem with the previous attempt was
that it could, in some situations, become hard to know which groups were
applied.

To get around this I decided to change things so that in the following
ways:

* The global ("") group is now always created on initialisation and
given that it contains systematics is always applied to all distributions
in the lh.

* This change comes with some required extra logic:
  * When a distribution is added to the lh, the systematics it is
associated with are checked so that they don't apply to same systematic
twice and also checks the "" groups for the same reason.
  * When adding to the "" group the systematic is checked against all
existing groups to make sure that can't be applied twice.
  * In this way it is not possible for systematics to be applied twice.
  * There is then logic to apply the "" group before any other groups
associated with that distribution if it is populated.

* To get the correct ordering when calculating the total response matrix
for a group I changed the operator used. *Problem for anyone using
systematics in their fits*.

* The new scheme doesn't allow for a "group chain" applied to a
distribution to contain the "" group. e.g.
{"alphaEnergyRes","","somethingElse"}. Something like this can easily be
achieved by specifying the groups explicitly.
  * Also the person has to know about the "" group, which can only be
known by looking the code or this commit message. And therefore they are
being too clever for me and deserve everything they get. But honestly I
tried to put that in and it is so hard in this scheme.

### CODE REVIEW

File: src/teststat/BinnedNLLH.cpp [hunk 1]
```diff
@@ -125,18 +149,19 @@ BinnedNLLH::GetBufferAsOverflow() const{
 }
 
 void
-BinnedNLLH::AddPdfs(const std::vector<BinnedED>& pdfs_){
```
Comment (terezakroupa) [inline src/teststat/BinnedNLLH.cpp:128]: I think this method should be retained for backward compatibility

File: src/teststat/BinnedNLLH.cpp [hunk 2]
```diff
@@ -62,25 +62,49 @@ BinnedNLLH::BinData(){
...
+BinnedNLLH::AddDist(const BinnedED& pdf_, const std::vector<std::string>& syss_){
+    fPdfManager.AddPdf(pdf_);
+    fSystematicManager.AddDist(pdf_,syss_);
 }
 
 void
-BinnedNLLH::AddPdf(const BinnedED& pdf_){
+BinnedNLLH::AddDist(const BinnedED& pdf_){
     fPdfManager.AddPdf(pdf_);
+    fSystematicManager.AddDist(pdf_,"");
```
Comment (terezakroupa) [inline src/teststat/BinnedNLLH.cpp:81]: Not sure why not just add this line to AddPdf(). Similarly, why not call AddDist(const BinnedED& pdf_, const std::vector<std::string>& syss_)  above (line 73) also AddPdf(...) etc.

**Review by terezakroupa — APPROVED** [review]

**Comment by terezakroupa** [comment]
Let's do the same here as with PR #2 and merge and then address the review comments in a separate PR. I'll add a merge-soon tag now and resolve the simple conflict when merging at the end of the day or tomorrow morning.

**Comment by tsengj10** [comment]
Do you have the "Resolve conflicts" button?  I think github will allow you to commit changes to the PR with this, so you don't have a merge conflict.

**Comment by terezakroupa** [comment]
@tsengj10 Yes, I do. I tried that, but if I want the resolution to be saved it says "commit merge", which seemed like might merge the entire PR. Upon more googling that seems like it's not the case, so I'll do that now.

**Comment by tsengj10** [comment]
looks fine. fine to merge.

File: test/unit/SystematicManagerTest.cpp [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    // void   SetTransformationObs(const ObsSet&){;}
+
+    // ObsSet GetTransformationObs() const{return ObsSet(3);} 
+    //
+    // void   SetDistributionObs(const ObsSet&){;}
+    // ObsSet GetDistributionObs() const{return ObsSet(3);}
+
+    void Construct() {;}
+
+    void   SetParameter(const std::string& name_, double value) {;}
```
[the remark this answers is not in the document]
Comment (terezakroupa) [reply test/unit/SystematicManagerTest.cpp:35]: Ah, because Systematic is an abstract class, got it

**Pull request merged**